src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -226,15 +237,22 @@ protected int doWork() {
                     rec.setDuplicateReadFlag(false);
                 }
             }
-            recordInFileIndex++;
 
-            if (!this.REMOVE_DUPLICATES || !rec.getDuplicateReadFlag()) {
-                if (PROGRAM_RECORD_ID != null) {
-                    rec.setAttribute(SAMTag.PG.name(), chainedPgIds.get(rec.getStringAttribute(SAMTag.PG.name())));
-                }
-                out.addAlignment(rec);
-                progress.record(rec);
+            // Manage the flagging of optical duplicates
+            final boolean isOpticalDuplicate = recordInFileIndex == nextOpticalDuplicateIndex;
+            if (isOpticalDuplicate) {
+                if (this.OPTICAL_DUPLICATE_TAG_NAME != null) rec.setAttribute(this.OPTICAL_DUPLICATE_TAG_NAME, this.OPTICAL_DUPLICATE_TAG_VALUE);
+                nextOpticalDuplicateIndex = this.opticalDuplicateIndices.hasNext() ? this.opticalDuplicateIndices.next() : -1;
             }
+
+            // Output the record if desired and bump the record index
+            recordInFileIndex++;
+            if (this.REMOVE_DUPLICATES && rec.getDuplicateReadFlag()) continue;
+            if (this.REMOVE_OPTICAL_DUPLICATES && isOpticalDuplicate) continue;
```

Comment:
if you move the index bump before the if(isBoolean), you will be able to put this continue inside the if statement. it's slightly simpler since it will put all the optical stuff in the same place.